crates/bevy_ecs/src/system/query.rs

```diff
@@ -659,6 +738,99 @@ where
         }
     }
 
+    /// Returns the query results for the given array of [`Entity`].
+    ///
+    /// In case of a nonexisting entity, duplicate entities or mismatched component, a [`QueryEntityError`] is
+    /// returned instead.
+    ///
+    /// See [`Query::multiple_mut`] for the infallible equivalent.
+    #[inline]
+    pub fn get_multiple_mut<const N: usize>(
+        &'s mut self,
+        entities: [Entity; N],
+    ) -> Result<[<Q::Fetch as Fetch>::Item; N], QueryEntityError> {
+        for i in 0..N {
+            for j in 0..i {
+                if entities[i] == entities[j] {
+                    return Err(QueryEntityError::AliasedMutability(entities[i]));
+                }
+            }
+        }
+
+        let array_of_results = entities.map(|entity| {
+            // SAFETY: Entities are checked for uniqueness above,
+            // the scheduler ensure that we do not have conflicting world access,
+            // and we require &mut self to avoid any other simultaneous operations on this Query
+            unsafe {
+                self.state.get_unchecked_manual::<Q::Fetch>(
+                    self.world,
+                    entity,
+                    self.last_change_tick,
+                    self.change_tick,
+                )
+            }
+        });
+
+        // If any of the entities were not present, return an error
+        for result in &array_of_results {
+            if let Err(QueryEntityError::NoSuchEntity(entity)) = result {
+                return Err(QueryEntityError::NoSuchEntity(*entity));
+            }
+        }
+
+        // Since we have verified that all entities are present, we can safely unwrap
+        Ok(array_of_results.map(|result| result.unwrap()))
+    }
+
+    /// Returns the query items for the provided array of [`Entity`]
+    ///
+    /// See [`Query::get_multiple_mut`] for the [`Result`]-returning equivalent.
+    ///
+    /// # Examples
+    ///
+    /// ```rust, no_run
+    /// use bevy_ecs::prelude::*;
+    ///
+    /// #[derive(Component)]
+    /// struct Spring{
+    ///     connected_entities: [Entity; 2],
+    ///     strength: f32,
+    /// }
+    ///
+    /// #[derive(Component)]
+    /// struct Position {
+    ///     x: f32,
+    ///     y: f32,
+    /// }
+    ///
+    /// #[derive(Component)]
+    /// struct Force {
+    ///     x: f32,
+    ///     y: f32,
+    /// }
+    ///
+    /// fn spring_forces(spring_query: Query<&Spring>, mut mass_query: Query<(&Position, &mut Force)>){
```

Comment:
This test is failing because our use of reference syntax has confused rustc.

```
error[E0623]: lifetime mismatch
  --> src\system\query.rs:815:82
   |
24 | fn spring_forces(spring_query: Query<&Spring>, mut mass_query: Query<(&Position, &mut Force)>){
   |                                                                ------------------------------ this type is declared with multiple lifetimes...
...
27 |          let [(position_1, mut force_1), (position_2, mut force_2)] = mass_query.multiple_mut(spring.connected_entities);
   |                                                                                  ^^^^^^^^^^^^ ...but data with one lifetime flows into the other here
   ```
   
   There's not actually two different lifetimes: both of those types have the same lifetime: that of the Query.

Reply:
Well that's a pretty bad error message...... there are two lifetimes on `Query` so its not obvious to me that we are confusing rustc with `&Position` and `&mut Force`

Reply:
I can silence the error by joining the `Query` lifetimes:

```rust
   fn spring_forces<'w, 's>(spring_query: Query<&Spring>, mut mass_query: Query<'s,'s, (&Position, &mut Force)>){
```

But then that leads to:

```
24 | fn spring_forces<'w, 's>(spring_query: Query<&Spring>, mut mass_query: Query<'s,'s, (&Position, &mut Force)>){
   |                      -- lifetime `'s` defined here
...
27 |          let [(position_1, mut force_1), (position_2, mut force_2)] = mass_query.multiple_mut(spring.connected_entities);
   |                                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                                                                       |
   |                                                                       `mass_query` was mutably borrowed here in the previous iteration of the loop
   |                                                                       argument requires that `mass_query` is borrowed for `'s`
```